### PR TITLE
feat: Save detour creation snapshot to db

### PR DESCRIPTION
**Pull request by hannahpurcell**
Still need to write tests and format, but wanted to gauge this approach asap

- [x] Tests!

### CODE REVIEW

File: assets/src/hooks/useDetour.ts
Comment (firestack) [inline assets/src/hooks/useDetour.ts:35]: issue: I think the entire `persistedSnapshot` object needs to be saved? not just the `context`

File: lib/skate/detours/db/detour.ex
Comment (firestack) [inline lib/skate/detours/db/detour.ex:10]: question: would it be possible to use the table `id` instead of creating a new column? I think we should be able to configure the ID to be a UUID, and I think we should have the database create the ID and not the frontend.

File: lib/skate_web/controllers/detours_controller.ex
Comment (firestack) [inline lib/skate_web/controllers/detours_controller.ex:14]: nit: I'd name this `upsert_snapshot` since it inserts or creates.

File: lib/skate_web/controllers/detours_controller.ex
Comment (firestack) [inline lib/skate_web/controllers/detours_controller.ex:23]: thought: along the lines of my [comment](https://github.com/mbta/skate/pull/2729/files#r1704411158), I would suggest that if `uuid` is not in the `context`, then we insert and return the UUID back to the frontend for the state machine to save. What do you think

File: lib/skate/detours/detours.ex
Comment (firestack) [inline lib/skate/detours/detours.ex:81]: can we use [`upserts`](https://hexdocs.pm/ecto/constraints-and-upserts.html#upserts) instead of creating this manually?

File: lib/skate_web/controllers/detours_controller.ex
Comment (firestack) [inline lib/skate_web/controllers/detours_controller.ex:23]:
```suggestion
    send_resp(conn, :ok, "")
```

Comment (firestack) [reply lib/skate_web/controllers/detours_controller.ex:23]: I think we should do something like this though
```suggestion
    if is_nil(uuid) do
      send_resp(conn, :created, new_uuid)
    else
      send_resp(conn, :ok, "")
    end
```

Comment (firestack) [reply lib/skate_web/controllers/detours_controller.ex:23]: And we'd get `new_uuid` from the database upsert

File: assets/src/hooks/useDetour.ts
Comment (firestack) [inline assets/src/hooks/useDetour.ts:34]: suggestion: I think what this conditional is getting at is that we don't want to save the detour unless we're in certain states. This code is currently using `uuid` as a sentinel value/proxy for "should we save the detour?"

What if instead, we use a tag, maybe called `no-save`, and do `snapshot.hasTag("no-save")` as the guard for if we call `putDetourUpdate`? 

I'll admit I played around with that idea when I built the localStorage stuff, so I'd be happy to pair on it if that'd be helpful.

File: assets/src/hooks/useDetour.ts
Comment (hannahpurcell) [reply assets/src/hooks/useDetour.ts:36]: I might as well do it here 👍

File: lib/skate/detours/db/detour.ex
Comment (hannahpurcell) [reply lib/skate/detours/db/detour.ex:10]: Oh nvm! I see your subsequent comment explains this exact process. Yeah, makes sense to me

Comment (firestack) [reply lib/skate/detours/db/detour.ex:10]: that sounds correct.

File: lib/skate/detours/detours.ex
Comment (hannahpurcell) [reply lib/skate/detours/detours.ex:81]: Hmmm I'm not sure. After removing the uuid from the db and using the `id` column as the unique identifier, the uuid coming from the frontend is either null (indicating the detour hasn't yet been added to the db) or it's a reference to the id of an existing detour. 

If the uuid was always non-null from the frontend, I see I could use upsert like in [Skate.Settings.User line 51](https://github.com/mbta/skate/blob/hp/test-branch-saving-to-db/lib/skate/settings/user.ex#L51).

File: lib/skate_web/controllers/detours_controller.ex
Comment (hannahpurcell) [reply lib/skate_web/controllers/detours_controller.ex]: Did a variation of this: always return the UUID, but only sometimes save it in the state machine. In hindsight, unsure if I took the right approach?

Comment (hannahpurcell) [reply lib/skate_web/controllers/detours_controller.ex]: Switched to using `json(conn)`

File: assets/src/hooks/useDetour.ts
Comment (hannahpurcell) [reply assets/src/hooks/useDetour.ts]: Ohhh I forgot to apply this. Given my other updates, would you still recommend using a tag?

File: assets/src/api.ts
Comment (firestack) [inline assets/src/api.ts:458]: suggestion: Can we use `apiCallResult` instead? This would make the error state a lot more explicit.

I can post a code example later if needed.

File: assets/src/hooks/useDetour.ts
Comment (firestack) [inline assets/src/hooks/useDetour.ts:37]: I don't think you need this. _maybe_ the UUID,
```suggestion
        if (uuid) {
          send({ type: "detour.set.uuid", uuid })
        }
```

but if we should _always_ get a UUID, then it should always be possible to send it right?
```suggestion
    send({ type: "detour.set.uuid", uuid })
```

[(I just realized) Maybe we should change out the type, because I think `null` is representing an error state here?](https://github.com/mbta/skate/pull/2729/files#r1711449321)

```suggestion
        if (isOk(uuid)) {
          send({ type: "detour.set.uuid", uuid: uuid.ok })
        }
```

File: lib/skate_web/controllers/detours_controller.ex
Comment (firestack) [inline lib/skate_web/controllers/detours_controller.ex:24]: suggestion: would this work?
```suggestion
    {:ok, %Skate.Detours.Db.Detour{id: returned_uuid} = returned_detour} =
      Detours.update_or_create_detour_for_user(user_id, uuid, %{
        state: snapshot
      })
```

File: assets/src/api.ts
Comment (firestack) [reply assets/src/api.ts:458]:
```suggestion
  apiCallResult(
    `/api/detours/update_snapshot`,
    object({ data: string() }),
    any(),
    {
      method: "PUT",
      headers: {
        "Content-Type": "application/json",
        "x-csrf-token": getCsrfToken(),
      },
      body: JSON.stringify({ snapshot: snapshot }),
    }
  )
```

File: lib/skate/detours/detours.ex
Comment (firestack) [inline lib/skate/detours/detours.ex:88]: issue: is `author` respected here? It seems like a user could update a detour that doesn't belong to them by injecting whatever ID they want.

I'm looking at providing a suggestion for all of this though, so you don't need to jump on this issue immediately, just wanted to bring it to your attention, maybe we write tests for this.

File: assets/src/models/createDetourMachine.ts
Comment (hannahpurcell) [reply assets/src/models/createDetourMachine.ts:63]: Oh yep, totally!! 🧠

File: assets/src/api.ts
Comment (hannahpurcell) [reply assets/src/api.ts:458]: Gotcha. (I was following in the predominant pattern of the rest of the api functions.)

So, I see `apiCallResult` is a recent add, but I'm not sure what new-and-different thing it does as compared to the other call methods? I'm thinking `apiCall`, `checkedApiCall`, and `apiCallResult` also need comments, especially if we're trying to shift towards a newer function over time

File: lib/skate/detours/detours.ex
Comment (firestack) [reply lib/skate/detours/detours.ex:88]: I tested with this and found that a user can update a detour that does not belong to them

```elixir
    test "update_or_create_detour_for_user/3 should not update a detour made by another user" do
      detour = detour_fixture()
      other_user = insert(:user)

      {:ok, new_detour} = Detours.update_or_create_detour_for_user(
          other_user.id,
          detour.id,
          %{state: %{a: 1}}
        )

      assert detour.id == new_detour.id
      assert detour.author_id == new_detour.author_id
      assert detour.state == new_detour.state
    end
```

File: assets/src/hooks/useDetour.ts
Comment (hannahpurcell) [reply assets/src/hooks/useDetour.ts]: I think this leads to a loop? The snapshotSubscription will trigger whenever anything with the snapshot changes, then will try to update the detour, and the detour update will try to update the snapshot, triggering the subscription again. My goal with the `snap.matches` was to only permit the initial set ID change

File: lib/skate/detours/detours.ex
Comment (firestack) [reply lib/skate/detours/detours.ex:88]: Hmm yeah that's fair! but it does bring up the issue of permissions, roles and access control, because as is, this still allows someone who doesn't have the `detours-pilot` role to update a detour, if they knew the endpoint. I don't think this matters for our threat model, but maybe we need to figure out our threat model.

File: assets/src/hooks/useDetour.ts
Comment (hannahpurcell) [reply assets/src/hooks/useDetour.ts]: We discussed on a slack huddle and decided yeah, the `matches` is annoyingly needed. Refactored anyway because of different issues

File: lib/skate_web/controllers/detours_controller.ex
Comment (hannahpurcell) [reply lib/skate_web/controllers/detours_controller.ex]: You bet!

File: assets/src/hooks/useDetour.ts
Comment (hannahpurcell) [reply assets/src/hooks/useDetour.ts]: Actually! I'd rather not do that transition, because I'd also need to make sure that we read from the database instead of localStorage, and that just feels like more than this PR should do.

File: assets/src/hooks/useDetour.ts
Comment (hannahpurcell) [reply assets/src/hooks/useDetour.ts]: Gosh did I try to make this work. But since we moved the `putDetourUpdate` to the state machine, there seems to be NO way of incorporating `hasTag` with a guard, and I'm so so frustrated about it! My "solution" ends up being pretty janky to get the same effect 😒

Comment (firestack) [reply assets/src/hooks/useDetour.ts]: dang that sucks, there's even [unofficial suggestions that there _should_ be a `hasTag`](https://github.com/statelyai/xstate/discussions/4347#discussioncomment-7404932) guard 😭

Comment (hannahpurcell) [reply assets/src/hooks/useDetour.ts]: I knooooowwwww I saw that toooo ugh

File: assets/src/hooks/useDetour.ts
Comment (joshlarson) [inline assets/src/hooks/useDetour.ts:40]: What happens if `.send` is called even if `.can` returns false? I assumed it wouldn't do anything, but I may have made that up.

Comment (hannahpurcell) [reply assets/src/hooks/useDetour.ts:40]: 🤔 The `can` clause serves as a guard here, so the `send` cannot execute if the `can` returns false, right? Now I'm second guessing everything

File: assets/tests/components/detours/diversionPage.saveDetour.test.tsx
Comment (joshlarson) [inline assets/tests/components/detours/diversionPage.saveDetour.test.tsx:1]: Nice!

File: assets/src/hooks/useDetour.ts
Comment (joshlarson) [reply assets/src/hooks/useDetour.ts:40]: Well, I tried removing the `.can` check and it caused the tests to hang, so I think my guess about how `.send` works was wrong 😅

Comment (firestack) [reply assets/src/hooks/useDetour.ts:40]: if `.send` is called, even though it _can't_ be handled more than once(with our current machine), it seems to cause the `subscribe` callback to be called again with a new snapshot, which gets into a loop where it's constantly triggering `send`->`subscribe`, unfortunately you'd need to guard the `send` some other way because of that, but it is a little surprising that `.send` doesn't do a `.can` internally and prevent `subscribe` from being invoked (though maybe it makes sense because `subscribe` does not describe if you're subscribing to anything _other_ than just snapshot updates and what causes a snapshot update does not seem to be explicitly defined either.)

File: assets/src/models/createDetourMachine.ts
Comment (firestack) [inline assets/src/models/createDetourMachine.ts:76]: **praise**: Love this comment!

File: assets/tests/components/detours/diversionPage.saveDetour.test.tsx
Comment (firestack) [inline assets/tests/components/detours/diversionPage.saveDetour.test.tsx:111]:
```suggestion
  // We made an assumption that we'll never want to save detour edits in response to changing route/variant
```

File: assets/tests/components/detours/diversionPage.saveDetour.test.tsx
Comment (firestack) [inline assets/tests/components/detours/diversionPage.saveDetour.test.tsx:79]:
```suggestion
  // We made an assumption that we'll never want to save detour edits in response to changing route/variant
```

File: assets/tests/api.test.ts
Comment (firestack) [inline assets/tests/api.test.ts:1155]: question/suggestion(non-blocking): could this be written using jest asymmetric matchers?

This may be the wrong code, but something like this?
```ts
expect(jest.mocked(window.fetch))
	.toHaveBeenCalledWith([
		expect.anything(),
		expect.objectContaining({
			method: "PUT",
			headers: expect.objectContaining({
				"x-csrf-token": expect.any(String) // or expect.anything(), I dunno the right syntax for specifying the type
			})
		})
	])
```

Comment (firestack) [reply assets/tests/api.test.ts:1155]: https://jestjs.io/docs/expect#asymmetric-matchers

**Review by firestack — COMMENTED** [review]
Looks good! I'll leave the review up to @joshlarson and come and look at this later once I'm back online and approve if it's still up

File: assets/src/hooks/useDetour.ts
Comment (hannahpurcell) [reply assets/src/hooks/useDetour.ts:40]: Ohhh oh oh, I understand your initial question now, after Kayla's reply. Yeah, we did that exact thing and tried just `send`ing, but ran into issues

**Review by firestack — APPROVED** [review]
I'd have to leave test structure/organization and methodology review to @joshlarson but LGTM!

File: assets/src/api.ts
Comment (firestack) [reply assets/src/api.ts]: Documentation! https://github.com/mbta/skate/pull/2747

**Comment by hannahpurcell** [comment]
Merging now, but would take later feedback for iteration / improvement!

**Pull request merged**
